fix: Create `RoutingFormResponseService` to get field value from identifier

joeauyeung opened the pull request:
# Refactor routing form response handling and revert identifier changes

## Summary

This PR refactors how routing form responses are handled in the Salesforce CRM integration by introducing a new service layer, while reverting problematic changes to the identifier field that were causing E2E test failures.

**Key Changes:**
- **Reverted identifier field changes**: Removed the `identifier` field from `FormResponse` type and related routing form components to fix E2E test failures
- **New service architecture**: Created `RoutingFormResponseService` and `RoutingFormResponseRepository` to encapsulate routing form response logic
- **CrmService refactor**: Updated Salesforce CRM service to use the new service layer instead of direct database queries for form responses

The original approach of storing identifier in response data caused type inconsistencies across routing forms components. Instead, this PR implements a cleaner architecture where identifiers are retrieved from the form definition when needed.

## Review & Testing Checklist for Human

- [ ] **Test Salesforce CRM integration end-to-end** - Create a routing form, submit it, and verify Salesforce records are created correctly with form field values
- [ ] **Verify E2E tests pass completely** - The original issue was failing E2E tests; confirm all routing forms E2E tests now pass 
- [ ] **Check type safety** - Ensure no TypeScript errors and that the identifier field removal doesn't cause runtime issues
- [ ] **Test new service layer edge cases** - Verify the `RoutingFormResponseService` handles missing responses, invalid identifiers, and malformed data gracefully

**Recommended test plan:** Create a routing form with various field types, submit responses through different paths (direct form, router endpoint), then verify Salesforce integration can correctly extract field values using identifiers.

---

### Diagram

```mermaid
%%{ init : { "theme" : "default" }}%%
graph TD
    CrmService["packages/app-store/salesforce/<br/>lib/CrmService.ts"]:::major-edit
    ResponseService["packages/lib/server/service/<br/>routingForm/routingFormResponse.service.ts"]:::major-edit
    ResponseRepo["packages/lib/server/repository/<br/>routingFormResponse.repository.ts"]:::major-edit
    FormInputFields["packages/app-store/routing-forms/<br/>components/FormInputFields.tsx"]:::minor-edit
    TypesDef["packages/app-store/routing-forms/<br/>types/types.d.ts"]:::minor-edit
    TestFiles["Test files<br/>(handleResponse.test.ts,<br/>onFormSubmission.test.ts)"]:::minor-edit
    
    CrmService -->|"uses"| ResponseService
    ResponseService -->|"uses"| ResponseRepo
    ResponseRepo -->|"queries"| Database[("Routing Form<br/>Responses DB")]:::context
    FormInputFields -->|"creates"| Database
    TypesDef -->|"defines types for"| FormInputFields
    TestFiles -->|"tests"| ResponseService
    
    subgraph Legend
        L1[Major Edit]:::major-edit
        L2[Minor Edit]:::minor-edit  
        L3[Context/No Edit]:::context
    end
    
    classDef major-edit fill:#90EE90
    classDef minor-edit fill:#87CEEB
    classDef context fill:#FFFFFF
```

### Notes

- This work was done by Devin AI in session: https://app.devin.ai/sessions/448c9e7cff2e40238d25540588b9b6be
- Requested by: @joeauyeung
- The new service layer follows repository pattern and could be extended for other routing form operations in the future
- Type changes may require careful monitoring in production to ensure no unexpected runtime errors

joeauyeung committed:
Make identifier required

joeauyeung committed:
Fallback to null if identifier isn't present

joeauyeung committed:
Type fix

joeauyeung commented on packages/app-store/routing-forms/lib/getResponseToStore.ts:
```diff
@@ -19,6 +19,7 @@ export const getResponseToStore = ({
 
     response[field.id] = {
       label: field.label,
+      identifier: field?.identifier || null,
```
This was generating the records to store. The field `identifier` wasn't being passed. All form fields should have an `identifier`

joeauyeung commented on packages/app-store/routing-forms/types/types.d.ts:
```diff
@@ -22,7 +22,7 @@ export type FormResponse = Record<
   {
     value: number | string | string[];
     label: string;
-    identifier?: string;
```
Ensure this property exists

joeauyeung commented on packages/trpc/server/routers/viewer/routing-forms/response.schema.ts:
```diff
@@ -6,7 +6,7 @@ export const ZResponseInputSchema = z.object({
   response: z.record(
     z.object({
       label: z.string(),
-      identifier: z.string().optional(),
```
Ensure this property exists

hariombalhara replied:
I think we should still keep it optional in this PR and later make it nullable in a separate PR as it might cause failure for certain embeds that are already loaded.

joeauyeung committed:
Type fixes

joeauyeung committed:
Type fix

joeauyeung committed:
Merge branch 'main' into add-form-identifier-to-response

hariombalhara commented on packages/app-store/routing-forms/components/FormInputFields.tsx:
```diff
@@ -73,7 +73,7 @@ export default function FormInputFields(props: FormInputFieldsProps) {
                     [field.id]: {
                       label: field.label,
-                      identifier: field?.identifier,
+                      identifier: field?.identifier || null,
```
If we are relying on identifier to be always set then I think we should use https://github.com/calcom/cal.com/blob/add-teams-routing-forms-response-record-endpoint-back/packages/app-store/routing-forms/lib/getFieldIdentifier.ts#L3

hariombalhara reviewed (changes requested):
My concern is why it is required to store the identifier.

If at some place we need the identifier from the response, we could always query the _fields_ by field.id as that is always there. This is how it has been thought of initially atleast. 

Also, it won't fix existing responses and for those we would still need to derive identifier on the fly.

joeauyeung committed:
Create `RoutingFormResponseRepository`

joeauyeung committed:
Create `RoutingFormResponseService`

joeauyeung committed:
Use repsotiories to find form value

joeauyeung committed:
Merge branch 'main' into add-form-identifier-to-response

joeauyeung committed:
Delete console.logs

joeauyeung committed:
Undo change in `ZResponseInputSchema` schema

joeauyeung committed:
Type fix

joeauyeung committed:
Undo changes

devin-ai-integration[bot] committed:
fix: correct import path in RoutingFormResponseService to resolve runtime errors

- Change relative import path to use @calcom alias
- Prevents import resolution failures that cause app startup issues

Co-Authored-By: joe@cal.com <j.auyeung419@gmail.com>

joeauyeung committed:
Undo changes

joeauyeung committed:
Update type

joeauyeung committed:
Update typing

hariombalhara commented on packages/lib/server/service/routingForm/routingFormResponse.service.ts:
```diff
@@ -0,0 +1,73 @@
+
+  async findFieldValueByIdentifier(identifier: string) {
+    // Find the field id based on the identifier that's passed
+    const field = this.fields.find((field) => field.identifier === identifier);
```
I think field.identifier could be _undefined_ too.  We need to use this logic here https://github.com/calcom/cal.com/blob/add-teams-routing-forms-response-record-endpoint-back/packages/app-store/routing-forms/lib/getFieldIdentifier.ts#L3 to ensure identifier is always available.


Note: _identifier_ is only set if it is different from label.

hariombalhara commented on packages/lib/server/service/routingForm/routingFormResponse.service.ts:
```diff
@@ -0,0 +1,73 @@
+
+    if (!field) {
+      this.log.error(`Field with identifier ${identifier} not found`);
+      throw new Error(`Field with identifier ${identifier} not found`);
```
I feel we shouldn't throw the error here. We could just choose to return _null_ here maybe.

Throwing error causes the caller code to handle it in try-catch

hariombalhara reviewed (dismissed):
LGTM overall !! 

There is just one blocking change https://github.com/calcom/cal.com/pull/22396/files#r2210882220

Others are suggestions

joeauyeung committed:
Address feedback

joeauyeung committed:
Address feedback

emrysal committed:
chore: Provide a suggestion for pr 22396, new structure (#22491)

* chore: Provide a suggestion for pr 22396, new structure

* Refactor to create two create methods with bookingUid and id

* Use `createWithBookingUid`

* Extract routing form response parser to seperate util

* Added more and improved test cases

* Fix

---------

Co-authored-by: Joe Au-Yeung <j.auyeung419@gmail.com>

emrysal committed:
Merge branch 'main' into add-form-identifier-to-response

emrysal committed:
Factory included wrong calls

emrysal committed:
Fix test for findFieldValueByIdentifier

joeauyeung committed:
Add tests

joeauyeung committed:
Fix test

joeauyeung committed:
Fix test

emrysal reviewed (approved):
LGTM 👍
